### PR TITLE
[WIP] Flavour

### PR DESCRIPTION
**Pull request by coiax**
:cl: coiax
add: Nanotrasen reminds its employees that they have ALWAYS been able to taste. Anyone claiming that they've recently only just gained the ability to taste are probably Syndicate agents.
/:cl:

Ports tasting from Bay.
- [x] Copy over most of the reagents
- [x] Be able to generate a taste_description_message from reagents
- [x] Actually copy over the proc that lets you taste things
- [x] Work out how general foodstuffs get their flavour
- [ ] Copy/make up all the general food flavours
- [x] Work out why spawned foods have flavours, but crafted foods do not
- [ ] BALANCEMEMES

### CODE REVIEW

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm [hunk 1]
```diff
@@ -86,18 +93,21 @@
...
 	description = "Delicious juice made from watermelon."
 	color = "#863333" // rgb: 134, 51, 51
+	taste_description = "juicy watermelon"
 
 /datum/reagent/consumable/lemonjuice
 	name = "Lemon Juice"
 	id = "lemonjuice"
 	description = "This juice is VERY sour."
 	color = "#863333" // rgb: 175, 175, 0
+	taste_description = "sourness"
```
Comment (ShadowDeath6) [inline code/modules/reagents/chemistry/reagents/drink_reagents.dm]: >not making it taste like space dosh

**Comment by Repukan** [comment]
So what are some practical uses for this?

Not criticizing, genuinely curious

**Comment by Shapsy** [comment]
Some of these just seem kind of weird, like there's a snarky narrator inside your head that's describing this stuff, like how a vodka martini tastes "shaken, not stirred", hooch tastes like "pure resignation", and irish coffee tastes like "giving up on the day".

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm [hunk 2]
```diff
@@ -9,6 +9,7 @@
 	id = "orangejuice"
 	description = "Both delicious AND rich in Vitamin C, what more do you need?"
 	color = "#E78108" // rgb: 231, 129, 8
+	taste_description = "oranges"
```
Comment (ImFlavo) [inline code/modules/reagents/chemistry/reagents/drink_reagents.dm]: @optimumtact

Comment (ShadowDeath6) [inline code/modules/reagents/chemistry/reagents/drink_reagents.dm]: knew somebody would do it

**Comment by Haevacht** [comment]
Repu I imagine detecting poisons in food would be a use for this.

**Comment by Repukan** [comment]
I suppose, aside from metacliques I don't see players eating often.

**Comment by Iamgoofball** [comment]
# I veto this code as only maintainer of chemistry

**Comment by RemieRichards** [comment]
@Iamgoofball opinion discarded.

**Comment by Haevacht** [comment]
Wait, he maintained goofchem? I thought that was the dominion of literally every other coder who was active at the time instead of him.

Anyway this is neat.

File: code/modules/reagents/chemistry/reagents/drink_reagents.dm [hunk 3]
```diff
@@ -462,68 +496,79 @@
...
 	description = "Juiced from real pumpkin."
 	color = "#FFA500"
+	taste_description = "pumpkin"
 
 /datum/reagent/consumable/blumpkinjuice
 	name = "Blumpkin Juice"
 	id = "blumpkinjuice"
 	description = "Juiced from real blumpkin."
 	color = "#00BFFF"
+	taste_description = "a mouthfull of pool water"
```
Comment (Militaires) [inline code/modules/reagents/chemistry/reagents/drink_reagents.dm]: U WOT M8

**Comment by Iamgoofball** [comment]
@RemieRichards No, opinion not discarded, this code is goofball-2013 tier badcode.

Fuck off.

**Comment by RemieRichards** [comment]
>it's shitcode
>it's literally just some variable definitions

Jesus christ goof it's a port, you can make improvements during ports you know.

**Comment by Iamgoofball** [comment]
# cute, i can't actually review it because he didn't add the garbage code

**Comment by XDTM** [comment]
# #largetext

**Comment by RemieRichards** [comment]
Goof I will discard any reviews you do, they're literally not relevant.

**Comment by dannno** [comment]
oh my fucking god goofball shut up

**Comment by Iamgoofball** [comment]
@RemieRichards im the only person in this codebase who gives a shit about chemistry code, don't give me that bullshit

i spent about 6 months working on this shit, don't fuck it up

throwing out my reviews because "lol goofball" is retarded and you should consider rethinking your strategy

should we just lock all PRs to maintainers only now if only maintainers are allowed to point out bad code

**Comment by Iamgoofball** [comment]
also re: its just variables

wait till you see the pile of shit that bay calls their chemistry taste code

**Comment by RemieRichards** [comment]
It's not that only maintainers can, it's that only people with standards can.

**Comment by Iamgoofball** [comment]
fuck you, asshole

**Comment by PKPenguin321** [comment]
Stop bullying each other

**Comment by jughu** [comment]
>tastes like acid
>you exactly know howe acid tastes
>why are you not dead

Cmon dont make it obvious like tastes like acid make it atleast less obvious

**Comment by XDTM** [comment]
>it tastes like pain and burning

**Comment by Iamgoofball** [comment]
Testing

**Comment by Iamgoofball** [comment]
>coiax blocked me so I can't review commits

File: code/modules/reagents/chemistry/holder.dm [hunk 4]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
 	var/list/cached_reagents = reagent_list
 	. = locate(type) in cached_reagents
 
+/datum/reagents/proc/generate_taste_message(mob/living/carbon/taster)
+	var/minimum_percent = 15
+
+	// TODO different species have different taste sensitivity?
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: Why? This'll get snowflake fast.

File: code/modules/reagents/chemistry/holder.dm [hunk 5]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
...
+	/*
+	if(ishuman(taster))
+		var/mob/living/carbon/human/H = taster
+		minimum_percent = round(15/ (H.isSynthetic() ? TASTE_DULL : H.species.taste_sensitivity))
+	*/
+
+	var/list/out = list()
+	var/list/tastes = list() //descriptor = strength
+	if(minimum_percent <= 100)
+		for(var/datum/reagent/R in reagent_list)
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: Typeless list

File: code/modules/reagents/chemistry/holder.dm [hunk 6]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
...
+					tastes[taste_desc] += taste_amount
+				else
+					tastes[taste_desc] = taste_amount
+
+		//deal with percentages
+		// TODO it would be great if we could sort these from strong to weak
+		var/total_taste = 0
+		for(var/taste_desc in tastes)
+			total_taste += tastes[taste_desc]
+		for(var/taste_desc in tastes)
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: >looping twice

File: code/modules/reagents/chemistry/holder.dm [hunk 7]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
...
+	*/
+
+	var/list/out = list()
+	var/list/tastes = list() //descriptor = strength
+	if(minimum_percent <= 100)
+		for(var/datum/reagent/R in reagent_list)
+			if(!R.taste_mult)
+				continue
+			if(istype(R, /datum/reagent/nutriment))
+				var/list/taste_data = R.data
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: >using data

Stop

File: code/modules/reagents/chemistry/holder.dm [hunk 8]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
...
+				for(var/taste in taste_data)
+					if(taste in tastes)
+						tastes[taste] += taste_data[taste]
+					else
+						tastes[taste] = taste_data[taste]
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: Oh god this is awful to read

File: code/modules/reagents/chemistry/holder.dm [hunk 9]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
...
+				else
+					tastes[taste_desc] = taste_amount
+
+		//deal with percentages
+		// TODO it would be great if we could sort these from strong to weak
+		var/total_taste = 0
+		for(var/taste_desc in tastes)
+			total_taste += tastes[taste_desc]
+		for(var/taste_desc in tastes)
+			var/percent = tastes[taste_desc]/total_taste * 100
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: Isn't tastedesc a string???

File: code/modules/reagents/chemistry/holder.dm [hunk 10]
```diff
@@ -685,6 +684,57 @@ var/const/INJECT = 5 //injection
...
+				continue
+			var/intensity_desc = "a hint of"
+			if(percent > minimum_percent * 2 || percent == 100)
+				intensity_desc = ""
+			else if(percent > minimum_percent * 3)
+				intensity_desc = "the strong flavor of"
+			if(intensity_desc != "")
+				out += "[intensity_desc] [taste_desc]"
+			else
+				out += "[taste_desc]"
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/holder.dm]: Can't wait for powergamers to speed search codebase for what poison is what

File: code/modules/reagents/chemistry/reagents/food_reagents.dm [hunk 11]
```diff
@@ -25,34 +27,55 @@
...
+
+	//cull all tastes below 10% of total
+	var/totalFlavor = 0
+	for(var/taste in data)
+		totalFlavor += data[taste]
+	if(!totalFlavor)
+		return
+	for(var/taste in data)
+		if(data[taste]/totalFlavor < 0.1)
+			data -= taste
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/reagents/food_reagents.dm]: Oh god no why stop using data why does this need done for all nutrient

File: code/modules/reagents/chemistry/reagents/food_reagents.dm [hunk 12]
```diff
@@ -25,34 +27,55 @@
...
+			data[taste] = newdata[taste]
+
+	//cull all tastes below 10% of total
+	var/totalFlavor = 0
+	for(var/taste in data)
+		totalFlavor += data[taste]
+	if(!totalFlavor)
+		return
+	for(var/taste in data)
+		if(data[taste]/totalFlavor < 0.1)
```
Comment (Iamgoofball) [inline code/modules/reagents/chemistry/reagents/food_reagents.dm]: AAAAAAAAAAAAAAA

**Review by Iamgoofball — CHANGES_REQUESTED** [review]
In the end this is some stupidly insane overhead for what amounts to little gain. How does this actually benefit things in it's current state? Ditch the get taste shit and make it on mob react via ingest or something.

File: code/modules/reagents/chemistry/reagents/pyrotechnic_reagents.dm [hunk 13]
```diff
@@ -5,6 +5,7 @@
 	description = "Thermite produces an aluminothermic reaction known as a thermite reaction. Can be used to melt walls."
 	reagent_state = SOLID
 	color = "#550000"
+	taste_description = "sweet tasting metal"
```
Comment (ShadowDeath6) [inline code/modules/reagents/chemistry/reagents/pyrotechnic_reagents.dm]: can it be reworded to like, "sweet and metallic"

**Comment by AdamElTablawy** [comment]
First off, Goofball, maybe people would listen to most of your valid points if you weren't a twat about it.

Same applies with Remie, except listen to goofball's valid points and don't be a twat.

Additionally, I don't see a point to making X poison have X taste especially since some poisons should logically be tasteless. It's a bit powergamey and dumb, also a nerf to poisoning food, rare enough as it is.

Finally, come on with the descriptions here for the bad shit. You have the chance to make CLF3 taste like 'OH JESUS CHRIST' in red text and you blew it. You fucking blew it. Come on.

**Comment by coiax** [comment]
@AdamElTablawy most of the poisons and medicines already taste like "bitterness". There is considerable overlap in most flavours. The most unique and distinctive ones are in alchohol.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm [hunk 14]
```diff
@@ -160,6 +164,7 @@
 	reagent_state = SOLID
 	color = "#669900" // rgb: 102, 153, 0
 	overdose_threshold = 30
+	taste_description = "fish"
```
Comment (AdamElTablawy) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: rezadone should taste like fish not cryoxadone, rezadone is made from carp

**Comment by PKPenguin321** [comment]
Kind of stupid. Remember how with poisons initially it would go "YOU FEEL LIKE YOU TOOK TOO MUCH OF POISON" in big red text with no subtlety and everybody hated it? Now it will do that for every single chem in the game even if you're not overdosing.

Seems like a bad change.

File: code/modules/reagents/chemistry/reagents/medicine_reagents.dm (hunk 14 above)
Comment (AdamElTablawy) [inline code/modules/reagents/chemistry/reagents/medicine_reagents.dm]: wait hold on nevermind i'm fucking retarded

**Comment by coiax** [comment]
Closing to open finished PR for actual review.

**Pull request closed without merging**